### PR TITLE
Add metadata verifier classes

### PR DESCRIPTION
**Pull request by tedbow**
Currently we do some verification of metadata in `Updater.php` like `checkFreezeAttack()` and some in metadata classes like `verifyNewHashes()`

We should create verified classes that are in charge of verifying metadata

Currently the only tests that should fail are in `\Tuf\Tests\Unit\Client\UpdaterTest` because this tested non-public methods. This could now test public methods on classes that extend `MetaDataVerifierBase`

### CODE REVIEW

File: src/Metadata/Verifier/VerifierBase.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+     *
+     * @param \Tuf\Client\SignatureVerifier $signatureVerifier
+     * @param \DateTimeImmutable $metadataExpiration
+     * @param \Tuf\Metadata\MetadataBase|null $trustedMetadata
+     */
+    public function __construct(SignatureVerifier $signatureVerifier, \DateTimeImmutable $metadataExpiration, ?MetadataBase $trustedMetadata = null)
+    {
+        $this->signatureVerifier = $signatureVerifier;
+        $this->metadataExpiration = $metadataExpiration;
+        if ($trustedMetadata && !$trustedMetadata->isTrusted()) {
```
Comment (tedbow) [inline src/Metadata/Verifier/VerifierBase.php]: We should use `\Tuf\Metadata\MetadataBase::ensureIsTrusted` here now that it is public

Comment (tedbow) [reply src/Metadata/Verifier/VerifierBase.php]: fixed

File: src/Metadata/Verifier/Factory.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+use Tuf\Metadata\Factory as MetadataFactory;
+use Tuf\Metadata\MetadataBase;
+use Tuf\Metadata\RootMetadata;
+use Tuf\Metadata\SnapshotMetadata;
+use Tuf\Metadata\TimestampMetadata;
+
+/**
+ * Defines a class that returns verifier objects for untrusted metadata.
+ */
+class Factory
```
[the remark this answers is not in the document]
Comment (phenaproxima) [reply src/Metadata/Verifier/Factory.php:15]: Yeah, `Verifier` would be more accurate, but then it conflicts with the `Verifier` sub-namespace. Need to think about this a bit...

File: src/Metadata/Verifier/VerifierBase.php [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        }
+        return $dateTime;
+    }
+
+    /**
+     * @param \Tuf\Metadata\MetadataBase $untrustedMetadata
+     */
+    protected function checkSignatures(MetadataBase $untrustedMetadata)
+    {
+        $this->signatureVerifier->checkSignatures($untrustedMetadata);
```
[the remark this answers is not in the document]
Comment (phenaproxima) [reply src/Metadata/Verifier/VerifierBase.php]: Agreed, and fixed.

File: src/Metadata/Verifier/Factory.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (tedbow) [reply src/Metadata/Verifier/Factory.php:15]: about `UniversalVerifier`

[the remark this answers is not in the document]
Comment (phenaproxima) [reply src/Metadata/Verifier/Factory.php]: Sure, I like that. Changed.

File: tests/Unit/Client/UpdaterTest.php [hunk 4]
```diff
@@ -38,22 +39,32 @@ public function testCheckRollbackAttackNoAttack(): void
         // We test lack of an exception in the positive test case.
         $this->expectNotToPerformAssertions();
 
-        // The incoming version is newer than the local version, so no
-        // rollback attack is present.
         $localMetadata = $this->getMockBuilder(MetadataBase::class)->disableOriginalConstructor()->getMock();
         $localMetadata->expects(self::any())->method('getType')->willReturn('any');
         $localMetadata->expects(self::any())->method('getVersion')->willReturn(1);
+
+        $verifier = new class ($localMetadata) extends VerifierBase
```
Comment (tedbow) [inline tests/Unit/Client/UpdaterTest.php:46]: All this should not live in UpdaterTest any more. We could just rename this class as it only tests methods that now are in verifier classes.

**Review by tedbow — COMMENTED** [review]
Just 1 remaining question

File: src/Metadata/Verifier/UniversalVerifier.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    /**
+     * Verifies an untrusted metadata object for a role.
+     *
+     * @param string $role
+     *   The metadata role (e.g. 'root', 'targets', etc.)
+     * @param \Tuf\Metadata\MetadataBase $untrustedMetadata
+     *   The untrusted metadata object.
+     *
+     * @throws \Tuf\Exception\PotentialAttackException\RollbackAttackException
```
[the remark this answers is not in the document]
Comment (phenaproxima) [reply src/Metadata/Verifier/UniversalVerifier.php:63]: I have no idea. IMHO we want to document all exceptions, but I also know that you and I have occasionally disagreed about exactly how and where to document exceptions :) I'm open to whatever you want to do here.

**Review by phenaproxima — APPROVED** [review]

**Pull request merged**